### ISSUE TITLE
[rllib] When using custom model with continuous action space, the num_outputs dimension doubles action_space dimension

### ISSUE THREAD

**Issue by longyuzhao** [ray-project/ray#17934]
### What is the problem?

When building a custom model with discrete action space, the num_outputs dimension equals to the action_space dimension. However, when building a custom model with continuous action space, the num_outputs dimension doubles the action_space dimension. This causes the action logits size doubles as well.

*Ray version and other system information (Python version, TensorFlow version, OS):*

ray==1.5.2
tensorflow==2.4.0

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

```
import gym
from gym import spaces
import numpy as np
import ray
import ray.rllib.agents.ppo as ppo
from ray.rllib.models import ModelCatalog
from ray.tune.registry import register_env
from ray.rllib.models.tf.tf_modelv2 import TFModelV2


class CustomEnv(gym.Env):
 
    def __init__(self, env_config={}):
        self.steps = 0

        self.action_space = spaces.Box(-1, 1, shape=(2,))
           
        self.dim = 3
 
        self.observation_space = spaces.Box(np.finfo(np.float32).min, np.finfo(np.float32).max, shape=(self.dim,))
           
    def step(self, action):
        self.steps += 1
        self.date_i += 1
           
        if self.date_i >= 50:
            done = True
        else:
            done = False
           
        state = np.array([1 for _ in range(self.dim)])
 
        reward = 0
        info = {'reward' : reward}
        return state, reward, done, info
 
    def reset(self):       
        self.date_i = 0
 
        state = np.array([1 for _ in range(self.dim)])
 
        return state

class CustomModel(TFModelV2):
    def __init__(self, obs_space, action_space, num_outputs, model_config, name):
        super(CustomModel, self).__init__(
            obs_space, action_space, num_outputs, model_config, name)
        
        print("obs space is {}".format(obs_space))
        print("action space is {}".format(action_space))
        print("num_outputs is {}".format(num_outputs))
        
ModelCatalog.register_custom_model('CustomModel', CustomModel)
register_env("TOYEnv", lambda env_config: CustomEnv(env_config))

ray.init(ignore_reinit_error=True)

trainer = ppo.PPOTrainer(env="TOYEnv", config={
    "model": {
        "custom_model": "CustomModel",
        # Extra kwargs to be passed to your model's c'tor.
        "custom_model_config": {},
    },
})
```

The output is:

```
(pid=15057) obs space is Box(-3.4028234663852886e+38, 3.4028234663852886e+38, (3,), float32)
(pid=15057) action space is Box(-1.0, 1.0, (2,), float32)
(pid=15057) num_outputs is 4
```

I don't understand why `num_outputs` is 4 not 2.

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by longyuzhao** [comment]
It seems that despite the `num_outputs` is doubled for the continuous actions, `PPOTrainer` can still output the correct action dimension. For example, for Pendulum-v0, `num_outputs` is 2 but `PPOTrainer` successfully outputs an action dimension of 1. I am not sure how this happens.

Also, I am trying to apply action mask to a hybrid action space (some continuous actions + some discrete actions). I want to only apply the action mask to the discrete actions. To do this, I want to apply an action mask to the action logits out of the custom model, and force the mask equal to 1 for those continuous actions (essentially doing nothing). However, if the dimension of action logits (which equals to `num_outputs`) is different from the action of action mask (which equals to `action_space`), I cannot add them together. Any suggestions? Thanks.

**Comment by zseymour** [comment]
@longyuzhao 
When you have a `Box` action space, it will be mapped to a `DiagGaussian` action distribution ([see here](https://github.com/ray-project/ray/blob/master/rllib/models/catalog.py#L262)).

Consequently, [this line](https://github.com/ray-project/ray/blob/master/rllib/models/tf/tf_action_dist.py#L264) splits the model output into two halves, such that the first half (2 elements in your case) parameterizes the distribution mean and the second half parameterizes the standard deviation.

For your masking issue, my recommendation/preference would be to compute logits for each subspace of your hybrid space separately, use the appropriate mask for the discrete subspaces, then concatenate all of the outputs at the end

**Comment by sven1977** [comment]
Hey @longyuzhao , thanks for posting this question. Closing this issue as @zseymour seemed to have given a good answer for this (thanks!). But feel free to re-open in case this does not solve your problem.
